src/util/thread_metrics.rs

```diff
@@ -78,62 +78,90 @@ impl Collector for ThreadsColletcor {
-        try!(fs::File::open(format!("/proc/{}/task/{}/comm", pid, tid))
-            .and_then(|mut f| f.read_to_string(&mut tname)));
-
-        pairs.push((tid, sanitize_thread_name(tid, tname)));
+        tids.push(tid);
     }
 
-    Ok(pairs)
+    Ok(tids)
 }
 
-fn sanitize_thread_name(tid: pid_t, tname: String) -> String {
-    let mut san = String::with_capacity(tname.len());
-    for c in tname.chars() {
+fn get_thread_name(tid: pid_t, stat: &str) -> (String, usize) {
+    let mut space_cnt = 0;
+    let mut open_brkt = 0;
+    let mut name = String::with_capacity(THD_NAME_LEN);
+    for c in stat.chars() {
         match c {
+            // TODO: properly handle `)a(`.
+            '(' => open_brkt += 1,
+            ')' => {
+                if open_brkt > 0 {
+                    open_brkt -= 1;
+                    if open_brkt == 0 {
+                        break;
+                    }
+                }
+            }
             // Prometheus label characters `[a-zA-Z0-9_:]`
             'a'...'z' | 'A'...'Z' | '0'...'9' | '_' | ':' => {
-                san.push(c);
+                if open_brkt > 0 {
+                    name.push(c);
+                }
+            }
+            '-' | ' ' => {
+                if open_brkt > 0 {
+                    if c == ' ' {
+                        space_cnt += 1;
+                    }
+                    name.push('_');
+                }
             }
-            '-' => san.push('_'),
             _ => (),
         }
     }
 
-    if san.is_empty() {
-        format!("{}", tid)
-    } else {
-        san
+    if name.is_empty() {
+        name = format!("{}", tid)
     }
+
+    (name, space_cnt)
 }
 
 fn to_err(s: String) -> Error {
     Error::new(ErrorKind::Other, s)
 }
 
+// Thread name's length is restricted to 16 characters,
+// including the terminating null byte ('\0').
+const THD_NAME_LEN: usize = 16;
+
 // See more man proc.
+// Index of utime and stime.
 const CPU_INDEX: [usize; 2] = [14 - 1, 15 - 1];
 
-fn find_thread_cpu_time(pid: pid_t, tid: pid_t) -> Result<(f64, f64)> {
-    let mut buffer = String::new();
+fn get_thread_stat(pid: pid_t, tid: pid_t) -> Result<(String, f64, f64)> {
+    let mut stat = String::new();
     try!(fs::File::open(format!("/proc/{}/task/{}/stat", pid, tid))
-        .and_then(|mut f| f.read_to_string(&mut buffer)));
+        .and_then(|mut f| f.read_to_string(&mut stat)));
+    Ok(get_thread_stat_internal(tid, &stat))
+}
 
-    let stats: Vec<_> = buffer.split_whitespace().collect();
-    let utime = stats[CPU_INDEX[0]];
-    let stime = stats[CPU_INDEX[1]];
-    Ok((utime.parse().unwrap(), stime.parse().unwrap()))
+// Extracted from `get_thread_stat`, for test purpose.
+fn get_thread_stat_internal(tid: pid_t, stat: &str) -> (String, f64, f64) {
+    let (name, space_cnt) = get_thread_name(tid, stat);
+    let stats: Vec<_> = stat.split_whitespace().collect();
+    let utime = stats[CPU_INDEX[0] + space_cnt];
```

Comment:
This is a little tricky. You can only handle the substring of stat which doesn't contains pid and pthread name.